src/software/geom/algorithms/intersects.cpp

```diff
@@ -165,3 +165,104 @@ bool intersects(const Segment &first, const Ray &second)
 {
     return intersects(second, first);
 }
+
+bool intersects(const Stadium &first, const Ray &second)
+{
+    auto start  = first.segment().getStart();
+    auto end    = first.segment().getEnd();
+    auto radius = first.radius();
+
+
+    // The parametric definition of a ray is r(d)=o+tu where o is the origin point and u
+    // is a unit vector Find the distance t along the ray that is closest to start and end
+    // Distance cannot be negative because rays only go positively along the unit vector u
+    auto startDist = std::fmax(0, second.toUnitVector().dot(start - second.getStart()));
+    auto endDist   = std::fmax(0, second.toUnitVector().dot(end - second.getStart()));
+
+    // Find corresponding point on the ray:
+    auto startClosestPoint = second.getStart() + second.toUnitVector() * startDist;
+    auto endClosestPoint   = second.getStart() + second.toUnitVector() * endDist;
+
+    // Check if square of distance is less than square of radius
+    // This will check if the ray is intersecting the circle with origin start
+    // and circle with origin end, all that is left is to check the two line segments
+    // connecting those circles
+    auto startIntersecting = distanceSquared(start, startClosestPoint) <= radius * radius;
+    auto endIntersecting   = distanceSquared(end, endClosestPoint) <= radius * radius;
+
+    Vector normal =
+        first.segment().toVector().rotate(Angle::fromDegrees(90)).normalize(radius);
+
+    auto s1 = first.segment() + normal;
+    auto s2 = first.segment() - normal;
+
+    return startIntersecting || endIntersecting || intersects(second, s1) ||
+           intersects(second, s2);
+}
+
+bool intersects(const Ray &first, const Stadium &second)
+{
+    return intersects(second, first);
+}
+
+bool intersects(const Stadium &first, const Circle &second)
+{
+    auto dist = distanceSquared(first.segment(), second.origin());
+
+    return dist <= std::pow(first.radius() + second.radius(), 2);
+}
+bool intersects(const Circle &first, const Stadium &second)
+{
+    return intersects(second, first);
+}
+
+bool intersects(const Stadium &first, const Segment &second)
+{
+    auto startDistance    = distance(first.segment(), second.getStart());
+    auto endDistance      = distance(first.segment(), second.getEnd());
+    auto startDistanceSeg = distance(first.segment().getStart(), second);
+    auto endDistanceSeg   = distance(first.segment().getEnd(), second);
+
+    auto shortestDistance =
+        std::fmin(startDistanceSeg,
+                  std::fmin(endDistanceSeg, std::fmin(startDistance, endDistance)));
```

Comment:
nit: you can call `std::min`/`std::minf` and pass it multiple values. E.g.
```suggestion
    double shortest_distance = std::min({start_distance, end_distance, start_distance_seg, end_distance_seg});
```